### PR TITLE
Add priority class to ztunnel

### PR DESCRIPTION
**Pull request by costinm**
**Please provide a description of this PR:**

Without this, ztunnel will not schedule (including for upgrades) on nodes that are full.
This change will kick out pods with lower (and default) priority.

Added a setting since in some clusters we may want to use system-node-priority - but since that
can be restricted by policies default to a custom (if lower) priority. We should probably do the same
in istio-cni and avoid running it in kube-system.

### CODE REVIEW

File: manifests/charts/ztunnel/templates/preemption.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+apiVersion: scheduling.k8s.io/v1
+kind: PriorityClass
```
Comment (howardjohn) [inline manifests/charts/ztunnel/templates/preemption.yaml:2]: Any prior art for doing this? Seems like other projects are just having users set it to system-node-critical and run in kube-system. Seems reasonable to do for us as well, I had assumed that is how ztunnel would be deployed

**Comment by costinm** [comment]
Not sure which projects do this - but even for istio-cni it seems wrong.

According to documentations ( or my understanding of the docs ) this is how
users can define classes and configure priority - and it doesn't seem right
for everyone
to just use system-node-critical and system-cluster-critical.

Plus it lets us keep using istio-system.

Seems to work - I don't see the downside really.

**Comment by costinm** [comment]
I don't mind system-node-critical - but I'm worried some K8S distro may put
restrictions, and I think it's a bad practice to lose whatever protection
namespace
provides and mix too many things in a namespace ( just like it's bad to run
gateways or grafana in istio-system ),

Ztunnel and istio-cni are probably a good fit for system-node-critical and
should be considered part of the CNI infra, but if that's the path we want
to take
we should start using kube-system in tests/etc ( it seems now we deploy in
istio-system by default )

**Comment by howardjohn** [comment]
Cilium, etc. I recall this was very intentional on istio-cni based on some
research that most platforms you cannot configure PriorityClass to begin
with.

I am fine with moving ztunnel to always kube-system by default

**Comment by costinm** [comment]
I don't mind moving ztunnel to kube-system by default. I assume it's not using any of the shared configs or secrets - we'll also need to check istio-system is not hidden in any of the constants and other defaults. It seems like a slightly larger change, and maybe we can do it separately.

But for istio-cni - we need to deal with upgrade/migrations, I think we shipped it in istio-system. I don't know what is the plan for this - and if we don't move it to kube-system we can't use the system-node-critical.

I don't like 2 ways of doing something - and ztunnel flip-flopping to different namespace depending on vendor is among the worse. Priority class explicitly defined seem like a standard and supported feature. 

Should we discuss it in the next meeting ? This PR at least allows ztunnel to have a priority class and it can be set to system-critical - and without it upgrades will be broken in pretty bad ways. Maybe merge this, and if we decide to move to kube-system revert ? Having different solution for cni and ztunnel also doesn't sound right.

**Comment by howardjohn** [comment]
Yes I don't think we should move CNI which already shipped, would be
problematic. On GKE we already document (and maybe auto-set, I forget) to
deploy in kube-system. Openshift has similar I believe.

Agree about the concerns...

**Comment by costinm** [comment]
But if we don't move CNI - and this mechanism works for both - we could as well keep both in istio-system.

Do you have any reference on why it wouldn't work on some platforms - but running in kube-system would ? The API seems intended for regular use - I would add a priority to istiod as well, because it is critical but not something we can move to kube-system.

It just feels wrong to run out stuff in kube-system ( and to even allow that...) - and just to run with higher priority when a CRD exists for this purpose.

**Comment by costinm** [comment]
Ok, how about this:
- change the PR to check if namespace is kube system, if yes use the higher
priority
- if not - still attempt to set the priority, but allow user to turn it off

Without any priority - it's really bad when upgrading.

I'll do the same for cni chart.

Downside: if a platform doesn't support priority, it would fail to deploy
but user can at least know and either use kube-system or run without a
priority and the risk of major outage if attempting to upgrade either on a
node that is full.

**Comment by howardjohn** [comment]
That is definitely "safest" (in terms of minimal change from current state,
maybe not safest for users) so I don't mind that. But I also am starting to
think I may have been wrong in my current comments, didn't get a chance to
properly research though

**Comment by costinm** [comment]
We can discuss it tomorrow in ambient meeting - I am ok with running ztunnel in either kube-system or istio-system, but 
we need to pick one, and ideally keep ztunnel and istio-cni consistent. 

I mentioned in another bug - if user ends up with istio-cni in both istio-system and kube-system, bad things are likely to
happen since they'll fight with each other in an infinite install loop.

**Comment by hzxuzhonghu** [comment]
Why would we run istio components in kube-system. That's not right, in some cluster, kube-system can not be written

**Comment by costinm** [comment]
I think the argument was that ztunnel is more or less part of the CNI infra
- and other CNI components are in kube-system.
And some vendors (like GKE) only allow the highest priority for kube-system
components.

I share the same concern - that some clusters may not allow running in
kube-system. But I also think CNI+Ztunnel are tightly coupled
with the real CNI and better off managed by the kube infra owner - AFAIK
the only safe upgrade path we have is coupled with
the node upgrades, in particular when our changes may interfere with
changes in upstream CNI ( with node upgrade it may fail,
but can be safely rolled back as a whole). So long term it may end up there
anyways.

**Pull request closed without merging**